src/logging.rs

```diff
@@ -0,0 +1,63 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use log::{Record, Metadata};
+use std::ffi::CString;
+use std::os::raw::c_char;
+use std::os::raw::c_int;
+
+pub type LogCallback = unsafe extern "C" fn(msg: *const c_char);
+
+struct LogProxy {level: log::Level, callback: LogCallback}
+
+impl log::Log for LogProxy {
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        metadata.level() <= self.level
+    }
+
+    fn log(&self, record: &Record) {
+        if self.enabled(record.metadata()) {
+            let msg = format!("{} - {}", record.level(), record.args());
+            // println!("{}", msg);  // Print the message to stdout
+            let cb: LogCallback = self.callback;
+            let s = CString::new(msg).unwrap();
+            let p = s.as_ptr();
+            unsafe {
+                cb(p);
+            }
+            // We do not use std::mem::forget(s), so the memory will be reclaimed
+            // by Rust once this function here. The callback should thus make a
+            // copy of the string. This is fine, because its probably converted to
+            // an internal (unicode) string type.
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+// Store the logger instance as a mutable nullable static object
+static mut LOGGER: Option<LogProxy> = None;
+
+#[no_mangle]
+pub unsafe extern "C" fn wgpu_set_log_callback(level: c_int, callback: LogCallback) {

```

Comment:
I would expect two API functions:
- `wgpu_set_log_callback(callback: LogCallback)`
- `wgpu_set_log_level(level: c_int)`

Reply:
Makes sense. Though combining them in a single function also works IMO. What do you think @kvark ?

Reply:
I'm trying to imagine why anyone would change the log level at run-time and I can't. @windelbouwman do you have a use case in mind that would require them to be separate functions?
It seems to me that a single one is good.

Reply:
I can imagine it's use in an interactive environment (like Python or Julia), where something does not work, you set the log level via the repl, and then try again to see what's going on.

Reply:
Also unit testing: Some tests could crank up the log level, do some stuff while capturing logs, and inspect the logs to see if things are looking good. Frankly, now that I've implemented it, and had a feel of it working, I prefer to have the two functions :)

Reply:
I have no use case, it's just that I think setting the callback and setting the level are two different things, and therefore I would expect two seperate API functions.

Reply:
two functions it is then!